benchmarks/be_loader.erl

```diff
@@ -0,0 +1,260 @@
+-module(be_loader).
+
+-behaviour(gen_server).
+
+%%
+%% Loads Boolean Expressions into BE-Tree
+%%
+
+-include("be_eval.hrl").
+
+%% API
+-export([
+  start_link/1,
+  stop/1,
+  load/2
+]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2,
+  code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+-record(be_loader_state, {
+  id :: atom(),
+  load_count :: integer()
+}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%% @doc Spawns the server and registers the local name (unique)
+-spec(start_link(Id :: atom()) ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link(Id) ->
+  gen_server:start_link({local, Id}, ?MODULE, [Id], []).
+
+%% @doc Stops the server
+stop(ServerRef) ->
+  gen_server:stop(ServerRef).
+
+%% @doc Loads Boolean Expressions from file
+load(ServerRef, FileName) ->
+  gen_server:call(ServerRef, {load, FileName}, infinity).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%% @private
+%% @doc Initializes the server
+-spec(init(Args :: term()) ->
+  {ok, State :: #be_loader_state{}} | {ok, State :: #be_loader_state{}, timeout() | hibernate} |
+  {stop, Reason :: term()} | ignore).
+init([Id]) ->
+  {ok, #be_loader_state{id = Id, load_count = 0}}.
+
+%% @private
+%% @doc Handling call messages
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
+    State :: #be_loader_state{}) ->
+  {reply, Reply :: term(), NewState :: #be_loader_state{}} |
+  {reply, Reply :: term(), NewState :: #be_loader_state{}, timeout() | hibernate} |
+  {noreply, NewState :: #be_loader_state{}} |
+  {noreply, NewState :: #be_loader_state{}, timeout() | hibernate} |
+  {stop, Reason :: term(), Reply :: term(), NewState :: #be_loader_state{}} |
+  {stop, Reason :: term(), NewState :: #be_loader_state{}}).
+handle_call({load, FileName}, _From, State = #be_loader_state{
+  id = Id, load_count = LoadCount}) ->
+  case be_bm_utils:file_exists(FileName) of
+    false ->
+      {reply, {error, {file_does_not_exist, FileName}}, State};
+    true ->
+      StrId = atom_to_list(Id),
+      LoadCount1 = LoadCount + 1,
+      StrCount = integer_to_list(LoadCount1),
+      ReaderId = list_to_atom(StrId ++ "_" ++ StrCount ++ "_loader_reader"),
```

Comment:
Atom leak?

Reply:
probably it is not critical, because it is only benchmark and it will not run in prod.

Reply:
Atoms leak addressed.